R/utilities-units.R

```diff
@@ -336,3 +336,212 @@ initializeDimensionAndUnitLists <- function() {
+  # function can handle them but not `split()`, which would drop the entire
+  # section of the data frame corresponding to `NA` and thus there will be loss
+  # of data when a data frame is split into a list of data frames.
+  #
+  # The trick is to create copies of source unit and molecular weight columns
+  # and fill in the missing values with something other than `NA`. Using these
+  # new columns with `split()` makes sure that the parts of a data frame where
+  # source units are missing won't be dropped. Note that the original columns
+  # containing source units remain unchanged.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    # Add suffix `Split` to the following columns:
+    # `xUnit`, `yUnit`, `yErrorUnit`, `molWeight`
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
+        .fns = as.character,
+        .names = "{.col}Split"
+      )
+    ) %>%
+    # Replace missing values in these new columns with `"missing"`, so that
+    # `split()` won't remove the corresponding portion of the data frame.
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Split$"),
+        .fns = ~ tidyr::replace_na(.x, "missing")
+      )
+    ) %>%
+    # `split()` will change the row order of the data frame depending on the
+    # alphabetical order of the levels of the variable used to split the data
+    # frame into a list.
+    #
+    # Therefore, an internal row identifier is kept to restore the original
+    # data frame row order before data is returned.
+    dplyr::mutate(.rowidInternal = dplyr::row_number())
+
+  # unit conversion --------------------------
+
+  # Split data frame to a list, mutate the unit column, and rebind.
+
+  # xUnit
+  xDataList <- .removeEmptyDataFrame(split(data, data$xUnitSplit))
+  data <- dplyr::bind_rows(lapply(xDataList, .xUnitConverter, xTargetUnit))
+
+  # yUnit
+  yDataList <- .removeEmptyDataFrame(split(data, list(data$yUnitSplit, data$molWeightSplit)))
+  data <- dplyr::bind_rows(lapply(yDataList, .yUnitConverter, yTargetUnit))
+
+  # yUnit error
+  if ("yErrorValues" %in% names(data)) {
+    yErrorDataList <- .removeEmptyDataFrame(split(data, list(data$yErrorUnitSplit, data$molWeightSplit)))
+    data <- dplyr::bind_rows(lapply(yErrorDataList, .yErrorUnitConverter, yTargetUnit))
+  }
+
+  # clean up and return --------------------------
+
+  # Restore the original row order using the internal row id
+  data <- dplyr::arrange(data, .rowidInternal)
+
+  # Remove all columns that were added only for internal workings of the function.
+  data <- dplyr::select(data, -dplyr::matches("Split$|.rowidInternal"))
+
+  return(data)
+}
+
+#' Remove empty data frames from a list
+#'
+#' @description
+#'
+#' Remove empty data frames sometimes produced due to the non-existent
+#' combination of source unit and molecular weight.
+#'
+#' @param x A list.
```

Comment:
so this is a list of dataframe?

Reply:
Yepp. 

Added that clarification, added an example, and created `.Rd` file for the same.